app/cmd/rpc/mesh/session.go

```diff
@@ -89,124 +94,497 @@ func (sn DispatchResponse) GetSupportedNodes() []string {
+		BlockHeight:  relay.Proof.SessionBlockHeight,
+		Nodes:        xsync.NewMapOf[*NodeSession](),
+		Dispatch:     nil,
+	}
+	session.Nodes.Store(relay.Proof.ServicerPubKey, &NodeSession{
+		PubKey:          relay.Proof.ServicerPubKey,
+		RemainingRelays: -1, // means that is unlimited until check it
+		RelayMeta:       &relay.Meta,
+		IsValid:         true, // true until node say the opposite
+		Validated:       false,
+		Error:           nil,
+		Session:         &session,
+	})
+
+	return &session
 }
 
-// getSessionHashFromRelay - calculate the session header and late the hash of it
-func getSessionHashFromRelay(r *pocketTypes.Relay) []byte {
-	header := pocketTypes.SessionHeader{
-		ApplicationPubKey:  r.Proof.Token.ApplicationPublicKey,
-		Chain:              r.Proof.Blockchain,
-		SessionBlockHeight: r.Proof.SessionBlockHeight,
+func (ss *SessionStorage) AddSessionToValidate(relay *pocketTypes.Relay) (*Session, error) {
+	sessionHeader := pocketTypes.SessionHeader{
+		ApplicationPubKey:  relay.Proof.Token.ApplicationPublicKey,
+		Chain:              relay.Proof.Blockchain,
+		SessionBlockHeight: relay.Proof.SessionBlockHeight,
 	}
 
-	return header.Hash()
+	var session *Session
+	hash := hex.EncodeToString(sessionHeader.Hash())
+
+	if v, ok := ss.Sessions.Load(hash); ok {
+		session = v
+		// add node to session if not there, but we already have the session
+		// this could happen because multiple nodes on the mesh are working for the same session,
+		// but the sessions are initialized by a relays, trusting on the incoming session retrieved.
+		// so each time a servicer require this session, is not in the session nodes list, it is added for then
+		// be sent to validate
+		if _, nodeOk := session.Nodes.Load(relay.Proof.ServicerPubKey); !nodeOk {
+			session.Nodes.Store(relay.Proof.ServicerPubKey, &NodeSession{
+				PubKey:          relay.Proof.ServicerPubKey,
+				RemainingRelays: -1, // means that is unlimited until check it
+				RelayMeta:       &relay.Meta,
+				Validated:       false, // mean this was not checked with fullnode yet
+				IsValid:         true,  // true until node say the opposite
+				Error:           nil,
+				Session:         session,
+			})
+		}
+	} else {
+		session = ss.NewSessionFromRelay(relay)
+		sessionStorage.Sessions.Store(hash, session)
+	}
+
+	// add this node/app/session relation to validate
+	ss.ValidationWorker.Submit(session.ValidateSessionTask(relay.Proof.ServicerPubKey))
+
+	return session, nil
+}
+
+func (ss *SessionStorage) ShouldAssumeOptimisticSession(relay *pocketTypes.Relay, servicerNode *fullNode) bool {
+	sessionBlockHeight := relay.Proof.SessionBlockHeight
+	fullNodeHeight := servicerNode.Status.Height
+	blocksPerSession := servicerNode.BlocksPerSession
+	return (sessionBlockHeight >= fullNodeHeight &&
+		(fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1)) &&
+		relay.Proof.SessionBlockHeight-servicerNode.GetLatestSessionBlockHeight() <= 1
```

Comment:
Assume session roll over:

1. Servicer on Height 9 (sessionBlockHeight 9)
2. Dispatcher on Height 8 (sessionBlockHeight 5)

the difference here should be <= blocksPerSession. We are bounding it by 1 session, correct?

Reply:
comments on the behavior of this function would be nice

Reply:
```fullNodeHeight%blocksPerSession == 0 || fullNodeHeight%blocksPerSession == 1```

wondering if this should be == 3, not <= 1

Reply:
@PoktBlade check it now, also uploaded the unit test for it.